Add Check Support for Java 21 Pattern Matching for Switch Syntax: BooleanExpressionComplexity

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation: https://checkstyle.org/checks/metrics/booleanexpressioncomplexity.html#BooleanExpressionComplexity
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---

```
➜  src cat config.xml                               
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="BooleanExpressionComplexity">

        </module>
    </module>
</module>
➜  src cat Test.java                                
record ColoredPoint(boolean p, boolean x, boolean c) {
}
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) {
}

public class Test {

    void test(Object obj) {
        switch (obj) {
            case ColoredPoint(boolean a, boolean b, boolean _)
                    when (a ^ (a || b) ^ (b || a) & (a | b)) -> {  // violation

                if (a ^ (a || b) ^ (b || a) & (a | b)) {
                }   // violation
                boolean c = (a ^ (a || b) ^ (b || a) & (a | b));   // violation

            }
            default -> System.out.println("none");
        }
    }

    public static void main(String[] args) {
        new Test().test(new ColoredPoint(false, false, true));
    }
}
➜  src javac --enable-preview --release 21 Test.java                        
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:11:21: Boolean expression complexity is 6 (max allowed is 3). [BooleanExpressionComplexity]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:13:17: Boolean expression complexity is 6 (max allowed is 3). [BooleanExpressionComplexity]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:15:27: Boolean expression complexity is 6 (max allowed is 3). [BooleanExpressionComplexity]
Audit done.
Checkstyle ends with 3 errors.


```

---

Describe what you want in detail

Check appears to work as intended after adding the nested EXPR token to WHEN expressions.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheckTest.java
@@ -133,4 +133,21 @@ public class BooleanExpressionComplexityCheckTest extends AbstractModuleTestSupp
                 expected);
     }
 
+    @Test
+    public void testWhenExpression() throws Exception {
+
+        final int max = 0;
+
+        final String[] expected = {
+            "17:21: " + getCheckMessage(MSG_KEY, 6, max),
+            "21:17: " + getCheckMessage(MSG_KEY, 6, max),
+            "25:27: " + getCheckMessage(MSG_KEY, 6, max),
+            "29:48: " + getCheckMessage(MSG_KEY, 1, max),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputBooleanExpressionComplexityWhenExpression.java"),
+                expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/booleanexpressioncomplexity/InputBooleanExpressionComplexityWhenExpression.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/booleanexpressioncomplexity/InputBooleanExpressionComplexityWhenExpression.java
@@ -1,0 +1,35 @@
+/*
+BooleanExpressionComplexity
+max = 0
+tokens = (default)LAND, BAND, LOR, BOR, BXOR
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.metrics.booleanexpressioncomplexity;
+
+public class InputBooleanExpressionComplexityWhenExpression {
+
+    void test(Object obj) {
+        switch (obj) {
+            case ColoredPoint(boolean a, boolean b, boolean _)
+                    when (a ^ (a || b) ^ (b || a) & (a | b)) -> {
+                // violation above, 'Boolean expression complexity is 6*.'
+
+                // violation below, 'Boolean expression complexity is 6*.'
+                if (a ^ (a || b) ^ (b || a) & (a | b)) {
+                }
+
+                // violation below, 'Boolean expression complexity is 6*.'
+                boolean c = (a ^ (a || b) ^ (b || a) & (a | b));
+
+            }
+            // violation below, 'Boolean expression complexity is 6*.'
+            case ColoredPoint(boolean a, _, _) when a && a -> { }
+            default -> System.out.println("none");
+        }
+    }
+
+    record ColoredPoint(boolean p, boolean x, boolean c) { }
+}

